File: python/Sort List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def sortList(self, head):
        if not head:
            return head
        cur, count = head, 0
        while cur:
            cur = cur.next
            count += 1
        return self.mergeSort(head, count)[0]

    def mergeSort(self, head, k):
        if k == 1:
            next = head.next
            head.next = None
            return head, next
        else:
            leftHead, next = self.mergeSort(head, k // 2)
            rightHead, next = self.mergeSort(next, k - (k // 2))
            return self.merge(leftHead, rightHead), next
    
    def merge(self, h1, h2):
        dummy = ListNode(0)
        cur = dummy
        while h1 and h2:
            if h1.val <= h2.val:
                cur.next = h1
                cur = cur.next
                h1 = h1.next
            else:
                cur.next = h2
                cur = cur.next
                h2 = h2.next
        if not h1 and h2:
            cur.next = h2
        elif not h2 and h1:
            cur.next = h1
        return dummy.next

File: cpp/Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (head == NULL) {
            return head;
        }
        int count = 0;
        ListNode *tmp = head;
        while (tmp) {
            tmp = tmp->next;
            count++;
        }
        return mergeSort(head, count);
    }
    
    ListNode* mergeSort(ListNode *head, int len) {
        if (len == 1) {
            head->next = NULL;
            return head;
        }
        ListNode *mid = head;
        for (int i = 0; i < len / 2; i++) {
            mid = mid->next;
        }
        ListNode *left = mergeSort(head, len / 2);
        ListNode *right = mergeSort(mid, len - (len / 2));
        return merge(left, right);
    }
    
    ListNode* merge(ListNode *leftHead, ListNode *rightHead) {
        ListNode *cursor, *head;
        if (leftHead && rightHead) {
            if (leftHead->val < rightHead->val) {
                cursor = leftHead;
                head = cursor;
                leftHead = leftHead->next;
            } else {
                cursor = rightHead;
                head = cursor;
                rightHead = rightHead->next;
            }
        } else if (leftHead == NULL) {
            return rightHead;
        } else if (rightHead == NULL) {
            return leftHead;
        }
        
        while (leftHead && rightHead) {
            if (leftHead->val < rightHead->val) {
                cursor->next = leftHead;
                cursor = cursor->next;
                leftHead = leftHead->next;
            } else {
                cursor->next = rightHead;
                cursor = cursor->next;
                rightHead = rightHead->next;
            }
        }
        if (leftHead == NULL) {
            cursor->next = rightHead;
        } else if (rightHead == NULL) {
            cursor->next = leftHead;
        }

        return head;
    }
};
